File: 백준/Silver/1010. 다리 놓기/다리 놓기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        StringBuilder sb = new StringBuilder();
        int[][] dp;
        
        int T = Integer.parseInt(br.readLine());
        for (int t = 1; t <= T; t++) {
        	st = new StringTokenizer(br.readLine());
        	int n = Integer.parseInt(st.nextToken());
        	int m = Integer.parseInt(st.nextToken());
        	dp = new int[m + 1][n + 1];
        	
        	for (int i = 1; i < m + 1; i++) {
        		for (int j = 0; j < n + 1; j++) {
        			if (j == i || j == 0) dp[i][j] = 1;
        			else dp[i][j] = dp[i - 1][j - 1] + dp[i - 1][j];
        		}
        	}
        	
        	sb.append(dp[m][n]).append("\n");
        }
        
        System.out.println(sb);
    }
}

File: 백준/Silver/1010. 다리 놓기/다리 놓기.kt
import java.util.*

val arr = Array(31) { IntArray(31) }

fun main() = with(System.`in`.bufferedReader()) {
    val t = readLine().toInt()
    val sb = StringBuilder()

    for (i in 1..30) {
        arr[1][i] = i
        arr[i][i] = 1
    }

    repeat(t) {
        val st = StringTokenizer(readLine())
        val n = st.nextToken().toInt()
        val m = st.nextToken().toInt()

        sb.appendLine(dp(n, m))
    }

    print(sb)
}

fun dp(n: Int, m: Int): Int {
    if (arr[n][m] != 0) return arr[n][m]
    arr[n][m] = dp(n - 1,m - 1) + dp(n, m - 1)
    return arr[n][m]
}